### PR TITLE
add option to fallback

### PR DESCRIPTION
**Pull request by mazlumtoprak**
This PR enables clients that boot from the azure server, to first boot from our own exposed server and then as fallback, use blobstorage.

### CODE REVIEW

**Comment by SwissGipfel** [comment]
General question regarding the netboot stack. Which files do we provide through Azure Storage Account and which files are synchronized to a netboot server?
We use `basicAuth` for the exposed network proxy but also provide every squashfs and kernel link with an additional `sas_token`. This confuses me.

File: netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env [hunk 1]
```diff
@@ -1 +1,7 @@
 NETBOOT_SERVER_IP="IP of the server the TFTP server is running on"
+AZURE_NETBOOT_SERVER_IP="IP of the Azure Netboot Server"
+AZURE_SYNC_SAS_TOKEN="SAS Token for fallback on Storage Account"
+AZURE_SYNC_BLOB_URL=https://insert-blob-name-here.blob.core.windows.net
+ONPREM_NETBOOT_SERVER_IP="IP of the OnPrem Netboot Server which is exposed"
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env:5]:
```suggestion
ONPREM_NETBOOT_SERVER_IP="IP of then on premise reverse proxy which is exposed"
```

File: netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env [hunk 2]
```diff
@@ -1 +1,7 @@
 NETBOOT_SERVER_IP="IP of the server the TFTP server is running on"
+AZURE_NETBOOT_SERVER_IP="IP of the Azure Netboot Server"
+AZURE_SYNC_SAS_TOKEN="SAS Token for fallback on Storage Account"
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env:3]: I thought, that the reverse proxy will be our fallback?

File: netboot-services/ipxeMenuGenerator/main.go [hunk 3]
```diff
@@ -33,10 +33,39 @@ func main() {
...
+		if azureNetbootServerIP == "" {
+			log.Fatal("AZURE_NETBOOT_SERVER_IP not set")
+		}
+
+		onpremExposedNetbootServer := os.Getenv("ONPREM_NETBOOT_SERVER")
+		if onpremExposedNetbootServer == "" {
+			log.Fatal("ONPREM_NETBOOT_SERVER not set")
+		}
+
+		azureBlobstorageSASToken := os.Getenv("AZURE_SYNC_SAS_TOKEN")
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/main.go:46]: Is the different behavior intended between `AZURE_SYNC_SAS_TOKEN` / `AZURE_SYNC_BLOB_URL` not set and the other environment variables. (`log.Error()` vs `log.Fatal()`)

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 [hunk 4]
```diff
@@ -5,16 +5,40 @@
...
 set language de_CH
 
+:check_boot_from_azure
+iseq ${next-server} {{ azureNetbootServerIP }} && goto register_basic_auth ||
+# if nothing applies, boot from local network
+set http-protocol http && set url ${next-server} && goto initial_menu ||
+
+:register_basic_auth
+set httpAuthUser {{ httpAuthUser }} && set httpAuthPassword {{ httpAuthPassword }} && goto check_if_onprem_netbootserver_is_reachable || 
+goto check_if_onprem_netbootserver_is_reachable ||
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/menu.ipxe.j2:25]: Why is this line needed?
This is the same `goto` as in line 24.

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 [hunk 5]
```diff
@@ -5,16 +5,40 @@
...
+
+:check_if_onprem_netbootserver_is_reachable
+imgfetch https://${httpAuthUser}:${httpAuthPassword}@{{ onpremExposedNetbootServer }}:8443/kernels/latest-kernel-version.json && goto set_onprem_netbootserver ||
+goto check_if_storage_account_is_reachable ||
+
+:check_if_storage_account_is_reachable
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/menu.ipxe.j2:31]: Do I confuse 2 different features now?
In my mind, this should be a check for the reverse proxy.

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 [hunk 6]
```diff
@@ -27,19 +51,20 @@ item --gap Erweitert:
 item advanced ${sp} Erweiterte Bootoptionen
 item reboot ${sp} Neustart
 item --gap Aktuell gesetzter Bootserver: ${next-server}
+item --gap Aktuell gesetzter Fileserver: {{ onpremExposedNetbootServer }}
```
Comment (SwissGipfel) [inline netboot-services/ipxeMenuGenerator/menu.ipxe.j2:54]: What's the difference between the `boot server` and the `file server`? Shouldn't it always be the netboot server, since the act as a cache for bigger locations?

File: netboot-services/ipxeMenuGenerator/advancedmenu.ipxe.j2 [hunk 7]
```diff
@@ -68,40 +66,30 @@ chain http://boot.netboot.xyz/ipxe/netboot.xyz-snponly.efi
 
 {% for file in prod %}
 :thinclient-{{ file.imageName }}
-set squash_url http://${next-server}/prod/{{ file.imageName }}.squashfs
-set kernel_url http://${next-server}/kernels/{{ file.kernelVersion }}/
+set squash_url ${http-protocol}://${basicAuth}${url}/prod/{{ file.imageName }}.squashfs${sas_token}
```
[the remark this answers is not in the document]
Comment (SwissGipfel) [reply netboot-services/ipxeMenuGenerator/advancedmenu.ipxe.j2:69]: Just saw, how the the variable is created. It's part of `basicAuth` because we don't always need basicAuth.

File: netboot-services/tftp/Dockerfile [hunk 8]
```diff
@@ -1,16 +1,23 @@
 FROM alpine:3.18 as bootfiles
-RUN apk add --no-cache bash gcc binutils make perl xz-libs xz-dev mtools git libc-dev
+RUN apk add --no-cache bash gcc binutils make perl xz-libs xz-dev mtools git libc-dev curl openssl coreutils
 WORKDIR /build
 # Clone iPXE
 RUN git clone git://git.ipxe.org/ipxe.git
 RUN sed -i 's/#undef\tDOWNLOAD_PROTO_HTTPS/#define\ DOWNLOAD_PROTO_HTTPS/' ipxe/src/config/general.h
+RUN sed -i 's/\/\/#define PING_CMD/#define PING_CMD/' ipxe/src/config/general.h
+RUN sed -i 's/\/\/#define NSLOOKUP_CMD/#define NSLOOKUP_CMD/' ipxe/src/config/general.h
+RUN export CERT=ca.pem,isrgrootx1.pem,lets-encrypt-r3.pem && export TRUST=ca.pem,isrgrootx1.pem,lets-encrypt-r3.pem
```
Comment (SwissGipfel) [inline netboot-services/tftp/Dockerfile:9]: I'd really like a inline Comment here, why we need to add let's-encrypt certificates. preferably with a link to the github issue, that you found. Because I assume, these cert chains are no longer needed once the github issue on the ipxe repository is fixed.

**Comment by mazlumtoprak** [comment]
> General question regarding the netboot stack. Which files do we provide through Azure Storage Account and which files are synchronized to a netboot server? We use `basicAuth` for the exposed network proxy but also provide every squashfs and kernel link with an additional `sas_token`. This confuses me.

the storage account and netboots server do have the same files available (except we want to opt in/out by declaring the `SYNC_DEV` / `SYNC_PROD` variables). The only difference between the storage account and netbootserver is the TFTP option and therefore the associated *.ipxe menu files. Those will be solely generated  and deliverd on a netbootserver. We can however still point to the storageaccount for downloading the squashfs and kernel files.

to clarify the usage of both `basicAuth` and `sas_token`: They are mutually exclusive. We only set the `sas_token`, when the `imgfetch` check on the onprem server fails. As on the jinja templating only populated variables are used, we can therefore assume that either one of those secrets is set, which saves us a good amount of code duplication.

That should also clear up on some other questions you had

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 (hunk 6 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/menu.ipxe.j2:54]: Good catch, this was on my part for debugging purposes and will be removed. To your question: on locations without a netboot server, it's possible that the TFTP Server is different from the server that provides us the squashfs files.

File: netboot-services/ipxeMenuGenerator/menu.ipxe.j2 (hunk 5 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/menu.ipxe.j2:31]: I struggled a bit with the naming here. The value is indeed resolving to the reverseproxy, which redirects traffic to our onprem netbootserver. I'm open for inputs

File: netboot-services/ipxeMenuGenerator/advancedmenu.ipxe.j2 (hunk 7 above)
[the remark this answers is not in the document]
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/advancedmenu.ipxe.j2:69]: described in my previous comment

File: netboot-services/tftp/Dockerfile (hunk 8 above)
Comment (mazlumtoprak) [reply netboot-services/tftp/Dockerfile:9]: added

File: netboot-services/ipxeMenuGenerator/main.go (hunk 3 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/main.go:46]: I wasn't sure if I should fatal here, if no azure details have been provided, as azure is the fallback option. I'm ok with it if we do check every variable to be filled out.

File: netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env (hunk 2 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env:3]: That wouldn't give us a big benefit as of now. We don't want to migrate the storage account to switzerland and the connection to azure from several locations is slow. First we want to capitalize on our own netbootserver and if it's unreachable, we go back to azure.

File: netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env (hunk 1 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/ipxe-menu-generator.env:5]: adjusted

File: netboot-services/ipxeMenuGenerator/main.go (hunk 3 above)
Comment (mazlumtoprak) [reply netboot-services/ipxeMenuGenerator/main.go:46]: Moved to fatal

**Review by SwissGipfel — APPROVED** [review]
